src/test/java/io/vertx/core/http/HttpBandwidthLimitingTest.java

```diff
@@ -0,0 +1,289 @@
+    Function<Vertx, HttpClient> http1ClientFactory = (v) -> v.createHttpClient();
+    Function<Vertx, HttpClient> http2ClientFactory = (v) -> v.createHttpClient(createHttp2ClientOptions());
+
+    return Arrays.asList(new Object[][] {
+      { 1.1, http1ServerFactory, http1ClientFactory },
+      { 2.0, http2ServerFactory, http2ClientFactory }
+    });
+  }
+
+  private Function<Vertx, HttpServer> serverFactory;
+  private Function<Vertx, HttpClient> clientFactory;
+
+  public HttpBandwidthLimitingTest(double protoVersion, Function<Vertx, HttpServer> serverFactory,
+                                   Function<Vertx, HttpClient> clientFactory) {
+    this.serverFactory = serverFactory;
+    this.clientFactory = clientFactory;
+  }
+
+  @Before
+  public void setUp() throws Exception {
+    super.setUp();
+    server = serverFactory.apply(vertx);
+    client = clientFactory.apply(vertx);
+  }
+
+  @After
+  public void after() throws InterruptedException
+  {
+    CountDownLatch waitForClose = new CountDownLatch(1);
+    vertx.close().onComplete(resp -> waitForClose.countDown());
+    awaitLatch(waitForClose);
+  }
+
+  @Test
+  public void sendBufferThrottled() throws Exception {
+    Buffer expectedBuffer = TestUtils.randomBuffer(TEST_CONTENT_SIZE);
+
+    HttpServer testServer = serverFactory.apply(vertx);
+    StopWatch watch = new StopWatch();
+    testServer.requestHandler(HANDLERS.bufferRead(expectedBuffer));
+    startServer(testServer);
+
+    watch.start();
+    HttpClient testClient = clientFactory.apply(vertx);
+    read(expectedBuffer, testServer, testClient);
+    await();
+    watch.stop();
+
+    assertTrue(watch.getTime() > expectedTimeMillis(TEST_CONTENT_SIZE, OUTBOUND_LIMIT));
+  }
+
+  @Test
+  public void sendFileIsThrottled() throws Exception {
+    StopWatch watch = new StopWatch();
+    HttpServer testServer = serverFactory.apply(vertx);
+    testServer.requestHandler(HANDLERS.getFile(sampleF));
+    startServer(testServer);
+
+    watch.start();
+    HttpClient testClient = clientFactory.apply(vertx);
+    testClient.request(HttpMethod.GET, testServer.actualPort(), DEFAULT_HTTP_HOST,"/get-file",
+                       req -> req.result().send(resp -> {
+                         resp.result().bodyHandler(r -> testComplete());
+                       }));
+    await();
+    watch.stop();
+
+    assertTrue(watch.getTime() > 1000);
+  }
+
+  @Test
+  public void dataUploadIsThrottled() throws Exception {
+    Buffer expectedBuffer = TestUtils.randomBuffer((TEST_CONTENT_SIZE));
+
+    StopWatch watch = new StopWatch();
+    HttpServer testServer = serverFactory.apply(vertx);
+    testServer.requestHandler(HANDLERS.bufferWrite(expectedBuffer));
+    startServer(testServer);
+
+    watch.start();
+    HttpClient testClient = clientFactory.apply(vertx);
+    write(expectedBuffer, testServer, testClient);
+    await();
+    watch.stop();
+
+    assertTrue(watch.getTime() > expectedTimeMillis(TEST_CONTENT_SIZE, INBOUND_LIMIT));
```

Comment:
we need here to use `Assert.assertTrue(...)` instead of `assert(...)` because `await()` was called before and the async test supposes the test is finished and no error can be reported